include/uncertainty_planning_core/uncertainty_contact_planning.hpp

```diff
@@ -141,6 +140,100 @@ class UncertaintyPolicyExecutionResult
   }
 };
 
+#if UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 2
+struct ExecutionTimeLimit {
+  ExecutionTimeLimit(double _secs)
+    : secs(_secs), clock(std::make_shared<rclcpp::Clock>())
+  {
+  }
+
+  ExecutionTimeLimit(double _secs, rclcpp::Clock::SharedPtr _clock)
+    : secs(_secs), clock(_clock)
+  {
+  }
+
+  double secs;
+  rclcpp::Clock::SharedPtr clock;
+};
+
+std::ostream& operator<<(
+    std::ostream& strm, const ExecutionTimeLimit& time_limit)
+{
+  return strm << time_limit.secs;
+}
+#elif UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 1
+using ExecutionTimeLimit = double;
+#endif
+
+class ExecutionTimer {
+public:
+#if UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 2
+  inline ExecutionTimer(ExecutionTimeLimit time_limit)
+    : clock_(time_limit.clock), time_limit_(time_limit.secs)
+  {
+  }
+#elif UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 1
+  inline ExecutionTimer(ExecutionTimeLimit time_limit)
+    : time_limit_(time_limit)
+  {
+  }
+#endif
+
+  inline void start()
```

Comment:
Let's change these methods a bit
```
double Start()
{
  if (start_time_)
  {
    throw std::logic_error("Execution timer already started");
  }
  start_time_ = OwningMaybe<double>(CurrentTime());
  return start_time_.Value();
}

double Stop()
{
  if (stop_time_)
  {
    throw std::logic_error("Execution timer already stopped");
  }
  stop_time_ = OwningMaybe<double>(CurrentTime());
  return stop_time_.Value();
}

bool IsExpired() const
{
  if (start_time_)
  {
    const double elapsed_time = CurrentTime() - start_time_.Value();
    return (elapsed_time >= time_limit_.Seconds());
  }
  return false;
}

double TimeLimit() const { return time_limit_.Seconds(); }

OwningMaybe<double> StartTime() const { return start_time_; }

OwningMaybe<double> StopTime() const { return stop_time_; }

OwningMaybe<double> ExecutionDuration() const
{
  if (start_time_ && stop_time_)
  {
    return OwningMaybe<double>(stop_time_.Value() - start_time_.Value());
  }
  else
  {
    return OwningMaybe<double>();
  }
}

double CurrentTime() const
{
#if UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 2
  return time_limit_.Clock().now().seconds();
#elif UNCERTAINTY_PLANNING_CORE__SUPPORTED_ROS_VERSION == 1
  return ros::Time::now().toSec();
#endif
}
```

This will require `mutable ExecutionTimeLimit time_limit_;`

Reply:
See https://github.com/calderpg/uncertainty_planning_core/pull/3/commits/2d3a145375854e0caadd7204655a6f32761e023c. There's a special case to be made for a zero time limit to mean no time limit.